### PR TITLE
Add section on building a client

### PR DESCRIPTION
**Pull request by JoshOrndorff**
This PR starts to fill in the starter code for chapter 4: Client.

In this section we walk through the steps of writing a blockchain client that will follow and build upon a blockchain network as it evolves.

I don't plan to make this perfect before the PR is merged (the whole repo is still WIP) but I figure the PR is a good way to get some initial feedback.

### CODE REVIEW

File: src/c4_client/mod.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+{
+    // You are free to add fields here. Please document them as you add them.
+    _ph_data: PhantomData<(C, SM, FC)>,
+
+    //TODO Probably don't want to include all of these right off the bat.
+    // For now it is more of a brain dump.
+    transaction_pool: Vec<SM::Transition>,
+    // block_database: HashMap<Hash, Block>,
+    // state_database: HashMap<Hash, State>,
+    leaves: HashSet<Hash>,
```
Comment (coax1d) [inline src/c4_client/mod.rs:59]: What are the leaves in this case?

File: src/c4_client/p1_data_structure.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+impl<C: Consensus, SM: StateMachine> Block<C, SM> {
+    /// Returns a new valid genesis block. By convention this block has no extrinsics.
+    pub fn genesis(genesis_state: &SM::State) -> Self {
+        todo!("Exercise 5")
+    }
+
+    /// Create and return a valid child block.
+    pub fn child(&self, pre_state: &SM::State, extrinsics: Vec<u8>) -> Self {
```
Comment (coax1d) [inline src/c4_client/p1_data_structure.rs:58]: a bit confused on the pre_state term but i guess I get it.
 
 So you basically just hash the `extrinsics` and then the `pre_state` and fill the header with the rest of the fields and place the extrinsics as the `body` field of the `Block` struct?
 
 I guess I got a little confused on pre_state.

**Review by coax1d — COMMENTED** [review]
Looks solid it is really generic. I am a little confused on leaves being a part of the `FullClient` struct. But I guess this is just a brain dump because you have to get the database and tree in there somewhere.

File: src/c4_client/mod.rs (hunk 1 above)
Comment (JoshOrndorff) [reply src/c4_client/mod.rs]: To answer the question: The leaves are all of the tips of the blockchain tree. A block that doesn't yet have any children is a leaf. Only leaves are candidates to be the longest chain.

But also, I removed this from the starter code. This was part of my braindump, and I think it is cleaner now. And actually, I think you only need to explicitly track the leaves in some fork choice rules, so it definitely shouldn't go directly on the client struct.

File: src/c4_client/p1_data_structure.rs (hunk 2 above)
Comment (JoshOrndorff) [reply src/c4_client/p1_data_structure.rs:58]: For the extrinsics, yeah, you just hash them and put the hash in the header.

But for the state it is different. You are given a starting state (which is also the ending state of the parent block). Then you execute the transactions to get the new ending state. Then you hash that new ending state and put it in the header.

**Comment by JoshOrndorff** [comment]
This is far from perfect, but I want to have something on `main` just in case any students are ready to try something hard or maybe contribute back to the project.

One HK student already mentioned to me that he has been working on this ahead of time.

**Review by gabriele-0201 — COMMENTED** [review]
Wow, it looks super nice! It offers many opportunities for students to discover client behavior that is not clear at first. I really liked how you handled generics!

Certainly, tests are missing, and I have not implemented them yet, so I am not sure if it works perfectly as is. However, it is definitely a great starting point for a fast student who developed the first three chapters quickly

The only note I feel to say is to probably avoid implementing an `Iterator` over the transaction pool, at least for now. It would add more rust syntax's complexity, shifting the focus from how a transaction pool works to how to make it work nicely with rust

**Pull request merged**